Angular/Angular guidelines and best practices/Assets & caching.md

```diff
@@ -0,0 +1,172 @@
+
+The files from the `assets/` directory do not get fingerprinted. Because of this, it is important to set up the correct caching mechanism for serving these files or else the client might be using stale files. There are two solutions:
+
+1. Well, add fingerprinting to these files!
+2. Utilize the transfer protocol caching mechanism
+
+The first solution is actually a hack, just like fingerprinting JS and CSS chunks is a hack. Ideally, we want to rely on the transfer protocol to handle the caching, without the need for fingerprinting.
+
+In the world of Web, application files are served via the HTTP protocol 99.99% of the time. It makes sense to use the built-in mechanisms offered by HTTP to handle caching of static assets. You can read more about the different caching mechanisms available in HTTP [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching).
+
+For our use case of static assets serving, we must adopt cache re-validation using ETags. The short description of this method is that the Web server calculates the hash of the file being served and sends that hash to the client together with the file itself. The client can cache this file and when fetching the same file in the future, the client sends that hash value of the cached file in the request headers as well. Depending on whether the hashes match or not, the server can return 200 with new version of the file or 304 without any payload.
+
+This caching mechanism could also be applied to JS and CSS chunks in a way that would not require fingerprinting those files, but since Angular already fingerprints those files, we will leave it as-is.
+
+_One small note_: if you use relative paths in SCSS, for example for a background image, then that image will get fingerprinted. However, if you have an image element in your template, like `<img src="/assets/logo.png">`, the `logo.png` file will not be fingerprinted and you are at risk of the client using a stale version of the file! Please check out [Loading assets via SCSS](#loading-assets-via-scss) sub-chapters that covers this topic.
+
+**Rule #3**: Implement cache re-validation using ETags for all static assets!
+
+## Cache re-validation implementation
+
+Depending on how the application files are served, implementation details for setting up HTTP caching will be a bit different.
+
+We will cover some of the common HTTP serving technologies, including static file serving and Server-side rendering.
+
+### Static file serving
+
+If the application files are served statically, the static file server must be configured to serve and cache assets files correctly.
+
+#### Nginx example
+
+The Nginx configuration file for the site should be updated to enable cache re-validation for the `assets/` directory:
```

Comment:
We are handling this on CEERIS at the moment with the `cache-control: no-cache` header and our devops team already added that to the nginx [template configuration](https://github.com/infinum/devops-ansible/blob/master/roles/nginx-vhost/templates/single-page-angular.j2) for Angular SPA applications. ETags are turned on by default in nginx. no-cache will compare ETags and download the new versions if values don't match. I think this should be written here.

Reply:
I updated both the nginx and node examples to set no-cache header.